indra/llmessage/llpacketring.cpp

```diff
@@ -43,329 +43,301 @@
-{
-    mUseInThrottle = use_throttle;
+    return send_packet( socket,
+                        headered_send_buffer,
+                        data_size + SOCKS_HEADER_SIZE,
+                        LLProxy::getInstance()->getUDPProxy().getAddress(),
+                        LLProxy::getInstance()->getUDPProxy().getPort());
 }
 
-void LLPacketRing::setUseOutThrottle(const bool use_throttle)
+bool LLPacketRing::sendPacket(int socket, const char * datap, S32 data_size, LLHost host)
 {
-    mUseOutThrottle = use_throttle;
+    mActualBytesOut += data_size;
+    return send_packet_helper(socket, datap, data_size, host);
 }
 
-void LLPacketRing::setInBandwidth(const F32 bps)
+void LLPacketRing::dropPackets (U32 num_to_drop)
 {
-    mInThrottle.setRate(bps);
+    mPacketsToDrop += num_to_drop;
 }
 
-void LLPacketRing::setOutBandwidth(const F32 bps)
+void LLPacketRing::setDropPercentage (F32 percent_to_drop)
 {
-    mOutThrottle.setRate(bps);
+    mDropPercentage = percent_to_drop;
 }
-///////////////////////////////////////////////////////////
-S32 LLPacketRing::receiveFromRing (S32 socket, char *datap)
-{
 
-    if (mInThrottle.checkOverflow(0))
+bool LLPacketRing::computeDrop()
+{
+    bool drop= (mDropPercentage > 0.0f && (ll_frand(100.f) < mDropPercentage));
+    if (drop)
     {
-        // We don't have enough bandwidth, don't give them a packet.
-        return 0;
+        ++mPacketsToDrop;
     }
-
-    LLPacketBuffer *packetp = NULL;
-    if (mReceiveQueue.empty())
+    if (mPacketsToDrop > 0)
     {
-        // No packets on the queue, don't give them any.
-        return 0;
+        --mPacketsToDrop;
+        drop = true;
     }
-
-    S32 packet_size = 0;
-    packetp = mReceiveQueue.front();
-    mReceiveQueue.pop();
-    packet_size = packetp->getSize();
-    if (packetp->getData() != NULL)
-    {
-        memcpy(datap, packetp->getData(), packet_size); /*Flawfinder: ignore*/
-    }
-    // need to set sender IP/port!!
-    mLastSender = packetp->getHost();
-    mLastReceivingIF = packetp->getReceivingInterface();
-    delete packetp;
-
-    this->mInBufferLength -= packet_size;
-
-    // Adjust the throttle
-    mInThrottle.throttleOverflow(packet_size * 8.f);
-    return packet_size;
+    return drop;
 }
 
-///////////////////////////////////////////////////////////
-S32 LLPacketRing::receivePacket (S32 socket, char *datap)
+S32 LLPacketRing::receiveOrDropPacket(S32 socket, char *datap, bool drop)
 {
     S32 packet_size = 0;
 
-    // If using the throttle, simulate a limited size input buffer.
-    if (mUseInThrottle)
+    // pull straight from socket
+    if (LLProxy::isSOCKSProxyEnabled())
     {
-        bool done = false;
-
-        // push any current net packet (if any) onto delay ring
-        while (!done)
+        char buffer[NET_BUFFER_SIZE + SOCKS_HEADER_SIZE];   /* Flawfinder ignore */
+        packet_size = receive_packet(socket, static_cast<char*>(static_cast<void*>(buffer)));
```

Comment:
is this double cast nessesary?

Reply:
Thanks for pointing this out.  I inherited  that line from elsewhere when buffer had been initialized to `U8[]`.  I will fix it.